cargo-dist/src/build/mod.rs

```diff
@@ -0,0 +1,229 @@
+
+            // These are symbols we expected! Save the path.
+            bin_result.sym_paths.push(sym_path);
+        }
+    }
+
+    /// Assuming the build is now complete, process the binaries as needed
+    ///
+    /// Currently this is:
+    ///
+    /// * checking src_path was set by found_bin
+    /// * computing linkage for the binary
+    /// * copying the binary and symbols to their final homes
+    ///
+    /// In the future this may also include:
+    ///
+    /// * code signing / hashing
+    /// * stripping
+    pub fn process_bins(&self, dist: &DistGraph, manifest: &mut DistManifest) -> DistResult<()> {
+        let mut missing = vec![];
+        for (pkg_id, pkg) in &self.packages {
+            for (bin_name, result_bin) in &pkg.binaries {
+                // If the src_path is missing, everything is bad
+                let Some(src_path) = result_bin.src_path.as_deref() else {
+                    missing.push((pkg_id.to_owned(), bin_name.to_owned()));
+                    continue;
+                };
+                if !src_path.exists() {
+                    missing.push((pkg_id.to_owned(), bin_name.to_owned()));
+                    continue;
+                }
+                let bin = dist.binary(result_bin.idx);
+
+                // compute linkage for the binary
+                self.compute_linkage(dist, manifest, result_bin, &bin.target)?;
+
+                // copy files to their final homes
+                self.copy_assets(result_bin, bin)?;
+            }
+        }
+
+        // FIXME: properly bulk these together instead of just returning the first
+        #[allow(clippy::never_loop)]
+        for (pkg_name, bin_name) in missing {
+            return Err(DistError::MissingBinaries { pkg_name, bin_name });
+        }
+
+        Ok(())
+    }
+
+    // Compute the linkage info for this binary
+    fn compute_linkage(
+        &self,
+        dist: &DistGraph,
+        manifest: &mut DistManifest,
+        src: &ExpectedBinary,
+        target: &TargetTriple,
+    ) -> DistResult<()> {
+        let src_path = src
+            .src_path
+            .as_ref()
+            .expect("bin src_path should have been checked by caller");
+
+        // If we're faking it, don't run the linkage stuff
+        let linkage = if self.fake {
+            // FIXME: fake this more interestingly!
+            let mut linkage = cargo_dist_schema::Linkage::default();
+            linkage.other.insert(cargo_dist_schema::Library {
+                path: "fakelib".to_owned(),
+                source: None,
+            });
+            linkage
+        } else {
+            determine_linkage(src_path, target)?.to_schema()
+        };
+        let bin = dist.binary(src.idx);
+        manifest.assets.insert(
+            bin.id.clone(),
+            AssetInfo {
+                id: bin.id.clone(),
+                name: bin.name.clone(),
+                system: dist.system_id.clone(),
+                linkage: Some(linkage),
```

Comment:
Mapping this per-asset instead of in a big array is going to be a *lot* nicer going forward, imo